### PR TITLE
fix(components): [upload] fileList prop should sync uploadFiles

### LINKED ISSUE #6491: [Bug Report] [Component] [upload] The component Upload attributes file-list is not reactive
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.31`
- Element Plus Version: `2.0.5`
- Browser / OS: `Chrome latest`
- Build Tool: `Vite`

### Related Component

- `el-upload`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtdXBsb2FkXG4gICAgY2xhc3M9XCJ1cGxvYWQtZGVtb1wiXG4gICAgYWN0aW9uPVwiaHR0cHM6Ly9qc29ucGxhY2Vob2xkZXIudHlwaWNvZGUuY29tL3Bvc3RzL1wiXG4gICAgOm9uLXByZXZpZXc9XCJoYW5kbGVQcmV2aWV3XCJcbiAgICA6b24tcmVtb3ZlPVwiaGFuZGxlUmVtb3ZlXCJcbiAgICA6YmVmb3JlLXJlbW92ZT1cImJlZm9yZVJlbW92ZVwiXG4gICAgbXVsdGlwbGVcbiAgICA6bGltaXQ9XCIzXCJcbiAgICA6b24tZXhjZWVkPVwiaGFuZGxlRXhjZWVkXCJcbiAgICA6ZmlsZS1saXN0PVwiZmlsZUxpc3RcIlxuICA+XG4gICAgPGVsLWJ1dHRvbiB0eXBlPVwicHJpbWFyeVwiPkNsaWNrIHRvIHVwbG9hZDwvZWwtYnV0dG9uPlxuICAgIDx0ZW1wbGF0ZSAjdGlwPlxuICAgICAgPGRpdiBjbGFzcz1cImVsLXVwbG9hZF9fdGlwXCI+XG4gICAgICAgIGpwZy9wbmcgZmlsZXMgd2l0aCBhIHNpemUgbGVzcyB0aGFuIDUwMGtiXG4gICAgICA8L2Rpdj5cbiAgICA8L3RlbXBsYXRlPlxuICA8L2VsLXVwbG9hZD5cbjwvdGVtcGxhdGU+XG48c2NyaXB0IGxhbmc9XCJ0c1wiIHNldHVwPlxuaW1wb3J0IHsgcmVmIH0gZnJvbSAndnVlJ1xuaW1wb3J0IHsgRWxNZXNzYWdlLCBFbE1lc3NhZ2VCb3ggfSBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgdHlwZSB7IFVwbG9hZEZpbGUgfSBmcm9tICdlbGVtZW50LXBsdXMvZXMvY29tcG9uZW50cy91cGxvYWQvc3JjL3VwbG9hZC50eXBlJ1xuXG5pbnRlcmZhY2UgUmF3RmlsZSB7XG4gIG5hbWU6IHN0cmluZ1xuICB1cmw6IHN0cmluZ1xufVxuXG5jb25zdCBmaWxlTGlzdCA9IHJlZjxSYXdGaWxlW10+KFtcbiAge1xuICAgIG5hbWU6ICdmb29kLmpwZWcnLFxuICAgIHVybDogJ2h0dHBzOi8vZnVzczEwLmVsZW1lY2RuLmNvbS8zLzYzLzRlN2YzYTE1NDI5YmZkYTk5YmNlNDJhMThjZGQxanBlZy5qcGVnP2ltYWdlTW9ncjIvdGh1bWJuYWlsLzM2MHgzNjAvZm9ybWF0L3dlYnAvcXVhbGl0eS8xMDAnLFxuICB9LFxuICB7XG4gICAgbmFtZTogJ2Zvb2QyLmpwZWcnLFxuICAgIHVybDogJ2h0dHBzOi8vZnVzczEwLmVsZW1lY2RuLmNvbS8zLzYzLzRlN2YzYTE1NDI5YmZkYTk5YmNlNDJhMThjZGQxanBlZy5qcGVnP2ltYWdlTW9ncjIvdGh1bWJuYWlsLzM2MHgzNjAvZm9ybWF0L3dlYnAvcXVhbGl0eS8xMDAnLFxuICB9LFxuXSlcblxuY29uc3QgaGFuZGxlUmVtb3ZlID0gKGZpbGU6IFVwbG9hZEZpbGUsIGZpbGVMaXN0OiBVcGxvYWRGaWxlW10pID0+IHtcbiAgY29uc29sZS5sb2coZmlsZSwgZmlsZUxpc3QpXG59XG5jb25zdCBoYW5kbGVQcmV2aWV3ID0gKGZpbGU6IFVwbG9hZEZpbGUpID0+IHtcbiAgY29uc29sZS5sb2coZmlsZSlcbn1cbmNvbnN0IGhhbmRsZUV4Y2VlZCA9IChmaWxlczogRmlsZUxpc3QsIGZpbGVMaXN0OiBVcGxvYWRGaWxlW10pID0+IHtcbiAgRWxNZXNzYWdlLndhcm5pbmcoXG4gICAgYFRoZSBsaW1pdCBpcyAzLCB5b3Ugc2VsZWN0ZWQgJHtmaWxlcy5sZW5ndGh9IGZpbGVzIHRoaXMgdGltZSwgYWRkIHVwIHRvICR7XG4gICAgICBmaWxlcy5sZW5ndGggKyBmaWxlTGlzdC5sZW5ndGhcbiAgICB9IHRvdGFsbHlgXG4gIClcbn1cbmNvbnN0IGJlZm9yZVJlbW92ZSA9IChmaWxlOiBVcGxvYWRGaWxlLCBmaWxlTGlzdDogVXBsb2FkRmlsZVtdKSA9PiB7XG4gIHJldHVybiBFbE1lc3NhZ2VCb3guY29uZmlybShgQ2FuY2VsIHRoZSB0cmFuc2ZlcnQgb2YgJHtmaWxlLm5hbWV9ID9gKVxufVxuPC9zY3JpcHQ+XG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XCJpbXBvcnRzXCI6e319In0=)

### Steps to reproduce

<template>
  <el-upload
    class="upload-demo"
    action="https://jsonplaceholder.typicode.com/posts/"
    :on-preview="handlePreview"
    :on-remove="handleRemove"
    :before-remove="beforeRemove"
    multiple
    :limit="3"
    :on-exceed="handleExceed"
    :file-list="fileList"
  >
    <el-button type="primary">Click to upload</el-button>
    <template #tip>
      <div class="el-upload__tip">
        jpg/png files with a size less than 500kb
      </div>
    </template>
  </el-upload>
</template>
<script lang="ts" setup>
import { ref } from 'vue'
import { ElMessage, ElMessageBox } from 'element-plus'
import type { UploadFile } from 'element-plus/es/components/upload/src/upload.type'

interface RawFile {
  name: string
  url: string
}

const fileList = ref<RawFile[]>([
  {
    name: 'food.jpeg',
    url: 'https://fuss10.elemecdn.com/3/63/4e7f3a15429bfda99bce42a18cdd1jpeg.jpeg?imageMogr2/thumbnail/360x360/format/webp/quality/100',
  },
  {
    name: 'food2.jpeg',
    url: 'https://fuss10.elemecdn.com/3/63/4e7f3a15429bfda99bce42a18cdd1jpeg.jpeg?imageMogr2/thumbnail/360x360/format/webp/quality/100',
  },
])

const handleRemove = (file: UploadFile, fileList: UploadFile[]) => {
  console.log(file, fileList)
}
const handlePreview = (file: UploadFile) => {
  console.log(file)
}
const handleExceed = (files: FileList, fileList: UploadFile[]) => {
  ElMessage.warning(
    `The limit is 3, you selected ${files.length} files this time, add up to ${
      files.length + fileList.length
    } totally`
  )
}
const beforeRemove = (file: UploadFile, fileList: UploadFile[]) => {
  return ElMessageBox.confirm(`Cancel the transfert of ${file.name} ?`)
}
</script>

### What is Expected?

Show default file list 

### What is actually happening?

Nothing and modify fileList manually is not reactive

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by exherb**
Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

https://togithub.com/element-plus/element-plus/issues/6014 removed this sync logic, any reason?

### CODE REVIEW

**Comment by exherb** [comment]
got time to review this? @emojiiii

**Comment by sxzz** [comment]
Could you please provide a minimum reproduction link?

**Comment by JeremyWuuuuu** [comment]
@exherb Can we hold this for a minute because we have a PR #6508 for refactoring this, I will notify you once that was done! :bow:

**Review by sxzz — APPROVED** [review]

**Comment by sxzz** [comment]
PTAL @JeremyWuuuuu

**Review by jw-foss — APPROVED** [review]

**Comment by exherb** [comment]
sorry, just saw the replys. Thank you

File: packages/components/upload/src/use-handlers.ts [hunk 1]
```diff
@@ -166,11 +166,12 @@ export const useHandlers = (
     () => props.fileList,
     (fileList) => {
       for (const file of fileList) {
-        file.uid = genFileId()
+        file.uid ||= genFileId()
         file.status ||= 'success'
       }
+      uploadFiles.value = fileList as UploadFiles
```
Comment (exherb) [inline packages/components/upload/src/use-handlers.ts:172]: - Force type is not a good idea
- Why watch deep? The main usage case is restore previous uploaded files from api returns.

**Pull request merged**